### PR TITLE
Fix poste routes to render 404 (not found) on malformed urls instead of failing

### PR DESCRIPTION
**Pull request by aoby**
Added relevant unit tests, and cleaned up some existing ones for `poste_routes`

Errors in the wild:
OpenSSL::Cipher::CipherError in honeybadger [here](https://app.honeybadger.io/projects/34365/faults/32633013#notice-summary) and [here](https://app.honeybadger.io/projects/34365/faults/31092603#notice-summary)

### CODE REVIEW

File: lib/cdo/poste.rb [hunk 1]
```diff
@@ -23,7 +23,10 @@ def self.decrypt(encrypted)
   end
 
   def self.decrypt_id(encrypted)
-    return decrypt(encrypted).to_i
+    decrypt(encrypted).to_i
+  rescue OpenSSL::Cipher::CipherError, ArgumentError => e
+    CDO.log.warn "Unable to decrypt poste id: #{encrypted}. Error: #{e.message}"
```
Comment (aoby) [inline lib/cdo/poste.rb:31]: I'm not sure we even care about logging this. Thoughts?

Comment (ashercodeorg) [reply lib/cdo/poste.rb:31]: Not sure I have enough context to be certain. What triggers the exception? Could this be useful for debugging a hacking attempt?

Comment (ashercodeorg) [reply lib/cdo/poste.rb:31]: nit: Please document the method returns an integer or nil.

File: pegasus/routes/poste_routes.rb [hunk 2]
```diff
@@ -27,7 +31,9 @@
   only_for 'code.org'
   dont_cache
   delivery = DB[:poste_deliveries].where(id: Poste.decrypt_id(id)).first
-  id = DB[:poste_opens].insert(delivery_id: delivery[:id], created_ip: request.ip, created_at: DateTime.now) if delivery
+  pass unless delivery
```
Comment (ashercodeorg) [inline pegasus/routes/poste_routes.rb]: Any reason to do the DB lookup if `decrypt_id` is `nil`?
```
decrypted_id = decrypt_id(id)
pass unless decrypted_id
delivery = DB[:poste_deliveries].where(id: decrypted_id).first
```

File: pegasus/routes/poste_routes.rb [hunk 3]
```diff
@@ -27,7 +31,9 @@
...
+
+  id = DB[:poste_opens].insert(delivery_id: delivery[:id], created_ip: request.ip, created_at: DateTime.now)
```
Comment (ashercodeorg) [inline pegasus/routes/poste_routes.rb]: nit: For readability, I would do `s/id/poste_opens_id`, updating the `headers` appropriately.

File: lib/cdo/poste.rb (hunk 1 above)
Comment (aoby) [reply lib/cdo/poste.rb:31]: That is possible. I'm really not sure why we've been getting bad urls, and thus far in honeybadger I haven't been able to see the actual url or params. I'll leave in the warning so we get a record of these. Given that we won't be seeing honeybadger errors, we likely will ignore it.

Adding documentation re: return value

File: pegasus/routes/poste_routes.rb (hunk 2 above)
Comment (aoby) [reply pegasus/routes/poste_routes.rb]: Nope, good call. Will fix

**Comment by aoby** [comment]
Ugh, I need to figure out how to address this test failure. I can repro it locally only sometimes when I run all the tests. It appears some test cases are running in parallel and interfering with our assertions. It does not repro when I run the test case in isolation.

**Comment by aoby** [comment]
It looks like the sporadic failure was a conflict with between creating test data in `DB`, vs. calling `Poste2.find_or_create_url` which inserted a row into `POSTE_DB`. I'm not entirely sure why it was causing a conflict, but doing everything directly with `DB` fixes it. I also added more validation to the test.

**Pull request merged**